plugin/formatting.py

```diff
@@ -13,25 +17,52 @@
 
 
 def options_for_view(view: sublime.View) -> 'Dict[str, Any]':
-    return {
-        "tabSize": view.settings().get("tab_size", 4),
-        "insertSpaces": True
-    }
+    return {"tabSize": view.settings().get("tab_size", 4), "insertSpaces": True}
 
 
 def apply_response_to_view(response: 'Optional[List[dict]]', view: sublime.View) -> None:
     edits = list(parse_text_edit(change) for change in response) if response else []
     view.run_command('lsp_apply_document_edit', {'changes': edits})
 
 
+class FormatOnSaveListener(sublime_plugin.ViewEventListener):
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+
+    @classmethod
+    def is_applicable(cls, view_settings: dict) -> bool:
+        syntax = view_settings.get('syntax')
+        if syntax:
+            return is_supported_syntax(syntax, client_configs.all)
+        return False
+
+    def on_pre_save(self) -> None:
+        file_path = self.view.file_name()
+        if file_path:
+            if self.view.settings().get("lsp_format_on_save"):
+                client = client_from_session(session_for_view(self.view, 'documentFormattingProvider'))
+                if client:
+                    params = {
+                        "textDocument": {
+                            "uri": filename_to_uri(file_path)
+                        },
+                        "options": options_for_view(self.view)
+                    }
+                    request = Request.formatting(params)
+                    response = client.execute_request(request)
```

Comment:
When I make a bunch of changes and then quickly hit CTRL+S, the textDocument/didChange notification is sent too late. We should have an additional `global_events.publish("view.on_purge_changes", self.view)` just above this line to make sure that the server sees the most recent document. Consider `git cherry-pick`ing this commit that already does that: 65e979a

Reply:
Good one, I wonder if our `didSave` should also purge changes before being sent?

Reply:
Don't know if relevant but with `pyls` server, I often see issues not disappearing on first save after fixing. This happens with `pylint` linter which is one of many that `pyls` runs and that one only runs on save.